### ISSUE TITLE
syslog-ng when sending logs to disk/elasticsearch consumes all system memory

### ISSUE THREAD

**Issue by nbari** [syslog-ng/syslog-ng#3321]
Hello, sorry if a GitHub "BUG" issue is not the correct place to raise this, but I don't know if it could be related to version `3.27` or some misconfiguration from my side, I started to notice that using the configuration below, the system starts to consume all system memory until consuming the swap, in a 32GB/8core system.

What I did to try to consume less  memory, is to reduced the ` elasticsearch-http - workers(4)` to `1` and also refused the ` log_fifo_size(1000);` and added  the disk-buffer section:

            disk-buffer(
                disk-buf-size(1073741824)
                mem-buf-length(1000)
                reliable(no)
                dir("/tmp")
            )

Any ideas about how could I fix or optimize no matter if the logs are delayed (not at least not lost) ?

Thanks in advance.

# syslog-ng
## Version of syslog-ng
```
syslog-ng 3 (3.27.1)
Config version: 3.22
Installer-Version: 3.27.1
Revision:
Compile-Date: Jun  4 2020 04:45:42
Module-Directory: /usr/local/lib/syslog-ng
Module-Path: /usr/local/lib/syslog-ng
Include-Path: /usr/local/share/syslog-ng/include
Available-Modules: basicfuncs,cryptofuncs,tags-parser,secure-logging,http,map-value-pairs,hook-commands,afsocket,afuser,afprog,tfgetent,csvparser,dbparser,syslogformat,cef,affile,graphite,xml,timestamp,appmodel,stardate,pseudofile,system-source,add-contextual-data,examples,disk-buffer,confgen,json-plugin,linux-kmsg-format,afstomp,kvformat,azure-auth-header
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: off
Enable-TCP-Wrapper: off
Enable-Linux-Caps: off
Enable-Systemd: off
```

## Platform
```
FreeBSD 12.1-STABLE r361928  amd64
```

## Configuration
```
@version:3.27
@include "scl.conf"

options {
    chain_hostnames(yes);
    keep_hostname(no);
    log_fifo_size(1000);
    normalize_hostnames(yes);
    stats_freq(0);
    threaded(yes);
    time_reopen(10);
    ts_format(iso);
    use_dns(no);
    dns-cache(no);
};

# ----------------------------------------------------------------------------
# sources
# ----------------------------------------------------------------------------
source s_local {
    system();
    internal();
};

source s_network {
    syslog(
            transport("tls")
            max-connections(100)
            log-iw-size(30000)
            so_keepalive(yes)
            tls(
                key-file("/usr/local/etc/syslog-ng/syslog.key")
                cert-file("/usr/local/etc/syslog-ng/syslog.crt")
                ca-dir("/usr/local/etc/syslog-ng/ca.d/")
               )
          );
};

# ----------------------------------------------------------------------------
# destinations
# ----------------------------------------------------------------------------
destination d_logs {
    file(
            "/var/log/remote/${HOST}/${MONTH}/${DAY}/$FACILITY.log"
            create-dirs(yes)
        );
};

destination d_http {
    elasticsearch-http(
            tls (
                peer-verify(no)
            )
            url("https://elk.example.com:9200/_bulk")
            user("user")
            password("secret")
            persist-name("syslog")
            index("syslog-${YEAR}.${MONTH}.${DAY}")
            batch-lines(100)
            batch-timeout(10000)
            workers(1)
            template("$(format-json --scope rfc5424 --scope nv-pairs --key ISODATE)")
            time-zone("UTC")
            type("syslog")
            disk-buffer(
                disk-buf-size(1073741824)
                mem-buf-length(1000)
                reliable(no)
                dir("/tmp")
            )
            );
};

destination d_http_haproxy {
    elasticsearch-http(
            tls (
                peer-verify(no)
            )
            url("https://elk.example.com:9200/_bulk")
            user("user")
            password("secret")
            persist-name("haproxy")
            index("haproxy-${YEAR}.${MONTH}.${DAY}")
            batch-lines(100)
            batch-timeout(10000)
            workers(1)
            time-zone("UTC")
            type("haproxy")
            template("$(format-json --scope rfc5424 --scope nv-pairs --key ISODATE)")
            disk-buffer(
                disk-buf-size(1073741824)
                mem-buf-length(1000)
                reliable(no)
                dir("/tmp")
            )
            );
};

destination d_all { file ("/var/log/all.log"); };

# ----------------------------------------------------------------------------
# filters
# ----------------------------------------------------------------------------
filter f_cron    { facility(cron); };
filter f_no_cron { not facility(cron); };
filter f_warning { level(warning..emerg); };
filter f_haproxy { program("haproxy"); };

# ----------------------------------------------------------------------------
# log line needs to start with '@json:{"foo":"bar"}'
# ----------------------------------------------------------------------------
block parser mixed-json-parser() {
    channel {
        junction {
            channel {
                parser {
                    json-parser(marker("@json:"));
                };
                flags(final);
            };
            channel {
                flags(final);
            };
        };
    };
};

# ----------------------------------------------------------------------------
# HAProxy json input
# ----------------------------------------------------------------------------
log {
    source(s_network);
    filter(f_haproxy);
    parser { mixed-json-parser(); };
    destination(d_logs);
    destination(d_http_haproxy);
    flags(flow-control, final);
};

# ----------------------------------------------------------------------------
# source_network
# ----------------------------------------------------------------------------
log {
    source(s_network);
    destination(d_logs);
    destination(d_http);
    flags(flow-control);
};

# ----------------------------------------------------------------------------
# local logs
# ----------------------------------------------------------------------------
log {
    source(s_local);
    destination(d_all);
};

log {
    source(s_local);
    filter(f_cron); filter(f_warning);
    destination(d_http);
    flags(final);
};

log {
    source(s_local);
    filter(f_no_cron); filter(f_warning);
    destination(d_http);
};
```

**Comment by MrAnno** [comment]
Hi,

The following should be taken into account for this configuration (this is just an approximate and theoretical calculation):

<details>
  <summary>Click to expand!</summary>
  
- global `log-fifo-size(1000)` - each memory queue can hold up to 1000 non-flow-controlled messages.
- `s_network(log-iw-size(30000) max-connections(100))`: 300 flow-controlled messages can be held in memory for each connection.
- `d_logs("${HOST}")`: A new file is created for every new host,  each opening a new destination queue with a possible `log-fifo-size()` number of in-memory messages. Theoretically, there is no upper limit for memory usage due to the `$HOST` macro, but your log paths are flow-controlled, so `s_network`'s `log-iw-size()` will be the limit.
- `d_http` and `d_http_haproxy`: each worker has a separate queue, maximum `mem-buf-length(1000) * workers(1)` number of messages will be stored on memory.

</details>

With 64K message lengths (it's a huge overestimation), `64K * 30000` = ~1.83 GB of flow-controlled messages and `64K * (1000 + 1000 + 1000)` = ~187.5 MB non-flow-controlled messages can be kept in memory.

(I've only considered messages stored in memory, there are other issues, of course, that can cause  high RAM usage.)

Overall, these values in your configuration do not explain the high memory usage in my opinion.

The following questions might shed light on the problem:
- `disk-buffer(dir("/tmp"))` has caught my eye. Are you sure it's not a `tmpfs` or any other memory-based mount?
- Is this a new problem that you've started to experience after upgrading to v3.27? If so, which version did you use before 3.27?
- How many different `${HOST}`s do you have?
- The `internal()` source currently has an infinite queue, so it is possible to create message loop, filling up the internal source queue. Please check `sbin/syslog-ng-ctl stats` and search for the `global;internal_queue_length;;a;processed` counter when the mem-usage is high. Validating other queue-related stats counters can be helpful as well.
- Lastly, can you narrow down the config and run syslog-ng under `valgrind` to look for memory leaks?

**Comment by nbari** [comment]
Hi  @MrAnno thanks for your reply, I use before versions  <= `3.25`, I started to notice this problem mainly because after upgrading I notice that the process was not running anymore, and this mainly due to the exhaustion of resources. 

`/tmp` is not a `tmpfs`, currently I run like 30 hosts, but had before close to 50, also traffic coming from `haproxy` is now less than 50% (I use to log like 400msg/sec) currently is around 50msg/sec.

This is the current output from `syslog-ng-ctl stats`

```
SourceName;SourceId;SourceInstance;State;Type;Number
global;payload_reallocs;;a;processed;6868397
global;sdata_updates;;a;processed;6914948
src.syslog;s_network;afsocket_sd.(stream,AF_INET(0.0.0.0:6514));a;connections;12
global;scratch_buffers_bytes;;a;queued;1536
dst.http;d_http_haproxy#0;http,https://elk.example.com:9200/_bulk;a;dropped;0
dst.http;d_http_haproxy#0;http,https://elk.example.com:9200/_bulk;a;processed;6896525
dst.http;d_http_haproxy#0;http,https://elk.example.com:9200/_bulk;a;queued;0
dst.http;d_http_haproxy#0;http,https://elk.example.com:9200/_bulk;a;written;6896500
destination;d_http_haproxy;;a;processed;6896525
global;msg_clones;;a;processed;6862905
destination;d_logs;;a;processed;6914948
global;internal_queue_length;;a;processed;0
source;s_local;;a;processed;225
src.internal;s_local#2;;a;processed;99
src.internal;s_local#2;;a;stamp;1592254303
source;s_network;;a;processed;6914948
destination;d_all;;a;processed;225
center;;received;a;processed;6915173
destination;d_http;;a;processed;18423
dst.http;d_http#0;http,https://elk.example.com:9200/_bulk;a;dropped;0
dst.http;d_http#0;http,https://elk.example.com:9200/_bulk;a;processed;18423
dst.http;d_http#0;http,https://elk.example.com:9200/_bulk;a;queued;0
dst.http;d_http#0;http,https://elk.example.com:9200/_bulk;a;written;18414
global;scratch_buffers_count;;a;queued;4024384356597
center;;queued;a;processed;13830121
```

I notice that ```global;scratch_buffers_count;;a;queued;4024384356597``` starts low but keeps growing 

Any advice, hint about how to run syslog-ng under `valgrind` 

thanks

**Comment by gaborznagy** [comment]
Hi @nbari!

Sorry for the late answer!
Do you still have the memory consumption issue?
I've read the issue comments, and sadly I have no new tips how to debug it further.
Regarding valgrind, I would recommend to run syslog-ng in the foreground with valgrind. In the meantime you should stop running syslog-ng concurrently as a service.

I would suggest the following command: 
`valgrind --log-file=<log-file-path> --leak-check=full --show-leak-kinds=all --trace-children=yes /usr/sbin/syslog-ng -F`
Run syslog-ng for a while and make sure messages are going through (i.e. log paths shouldn't be idle). After that stop syslog-ng and collect the valgrind log file.

On a sidenote: the `global;scratch_buffers_count;;a;queued;` was a bug in statistics, there were no problem with the memory usage (fixed in #3355).

**Comment by nbari** [comment]
Hi @gaborznagy thanks, I will close this for now and come back if found something
